test/09-loki-backend.ts

```diff
@@ -0,0 +1,128 @@
+import _ = require('lodash');
+import * as Bluebird from 'bluebird';
+import { expect } from './test-lib/chai';
+import { LokiBackend } from '../src/features/device-logs/lib/backends/loki';
+
+const createLog = (extra = {}) => {
+	return {
+		isStdErr: true,
+		isSystem: true,
+		message: `a log line`,
+		timestamp: Date.now(),
+		createdAt: Date.now(),
+		...extra,
+	};
+};
+
+const createContext = (extra = {}) => {
+	return {
+		id: 1,
+		uuid: '1',
+		belongs_to__application: 1,
+		images: [],
+		...extra,
+	};
+};
+
+describe('loki backend', () => {
+	it('should successfully publish log', async () => {
+		const loki = new LokiBackend();
+		const response = await loki.publish(createContext(), [createLog()]);
+		expect(response).to.be.not.null;
+	});
+
+	it('should store and retrieve device log', async () => {
+		const loki = new LokiBackend();
+		const ctx = createContext();
+		const log = createLog();
+		const response = await loki.publish(ctx, [_.clone(log)]);
+		expect(response).to.be.not.null;
+		const history = await loki.history(ctx, 1000);
+		expect(history[history.length - 1]).to.deep.equal(log);
+	});
+
+	it('should convert multiple logs with different labels to streams and then back to logs', function () {
+		const loki = new LokiBackend();
+		const ctx = createContext();
+		const logs = [
+			createLog(),
+			createLog(),
+			createLog({ serviceId: 1 }),
+			createLog({ serviceId: 2 }),
+			createLog({ serviceId: 3 }),
+		];
+		// @ts-ignore usage of private function
+		const streams = loki.fromDeviceLogsToStreams(ctx, _.cloneDeep(logs));
+		expect(streams.length).to.be.equal(
+			4,
+			'should be 4 streams since 5 logs have 4 distinct services (null, 1, 2, 3)',
+		);
+		// @ts-ignore usage of private function
```

Comment:
Can you make this a
```suggestion
		// @ts-expect-error usage of private function
```
just for the slight extra safety, and same for the other instances of `ts-ignore` where we'd expect an error

Reply:
updated